### PR TITLE
[Merged by Bors] - Make webserver timeout configurable. Increase default to 5m

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Running into this with the Superset demos. Also had to increase it within my previous company. 60s is really low :/

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by sbernauer** [comment]
Hmm maybe make it overwrite-able :thinking: . But i honestly i don't see a benefit in decreasing it. So we could pick a very high value and hard-code it. In the case of whatever reason we can still add the complexity to make it user-configurable

**Comment by teozkr** [comment]
Yeah I think 6h is well into the territory of "this should be opt-in". If you lose 60s of work to a timeout you didn't know about, that's annoying. If you lose 6h of work to a timeout you didn't know about, well...

**Comment by maltesander** [comment]
> Yeah I think 6h is well into the territory of "this should be opt-in". If you lose 60s of work to a timeout you didn't know about, that's annoying. If you lose 6h of work to a timeout you didn't know about, well...

Yeah it feels like quite an important setting :D 
I would make it an option in the CRD, keep the 60s default (which is what people should be used to / know about) and mention it in the docs / examples.

**Comment by sbernauer** [comment]
The problem i see here is that if we default to 60s and querying e.g. a Trino or Druid with more than 10 rows, every user will run into this problem. For me personally a Superset that times out every query after 60s is unusable. You can get a sense of felling in the demo from https://github.com/stackabletech/stackablectl/pull/66.
And I don't like the idea that 90% of the users run into a problem and have to manually increase the timeout.
Can you please elaborate on the "If you lose 60s of work to a timeout you didn't know about, that's annoying. If you lose 6h of work to a timeout you didn't know about, well..." as I'm trying to understand how this work loss can happen. Especially work loss that could have been prevented by setting the timeout to 10m vs 6h.

**Comment by lfrancke** [comment]
Just for reference: https://superset.apache.org/docs/frequently-asked-questions/#why-are-my-queries-timing-out https://github.com/apache/superset/issues/5993

**Comment by maltesander** [comment]
> The problem i see here is that if we default to 60s and querying e.g. a Trino or Druid with more than 10 rows, every user will run into this problem. For me personally a Superset that times out every query after 60s is unusable. You can get a sense of felling in the demo from [stackabletech/stackablectl#66](https://github.com/stackabletech/stackablectl/pull/66). And I don't like the idea that 90% of the users run into a problem and have to manually increase the timeout. Can you please elaborate on the "If you lose 60s of work to a timeout you didn't know about, that's annoying. If you lose 6h of work to a timeout you didn't know about, well..." as I'm trying to understand how this work loss can happen. Especially work loss that could have been prevented by setting the timeout to 10m vs 6h.

If you default the timeout to 6h, and 6h is not enough you lose 6h. If you keep the default on 60s and it fails you lose 60s which is fine for the user to check what happened and increase it himself. 

I dont really care about the default as long as its documented. But i still think this is a must have configuration setting.

**Comment by sbernauer** [comment]
Started test https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/superset-operator-it-custom/5/console

**Comment by teozkr** [comment]
Just to make sure I understand this… this is the timeout for queries that run on either page load?

**Comment by sbernauer** [comment]
Tests were green.
I don't have any additional information other than https://superset.apache.org/docs/frequently-asked-questions/#why-are-my-queries-timing-out as well as the information that queries taking longer than 60s (which is common on Trino) will let the Superset Dashboard or Chart display a Timeout message besides the query continuing running and finish in the background.
This PR fixes that problem. Be need both - `SUPERSET_WEBSERVER_TIMEOUT` and the `--timeout` to the start-command to solve it

File: rust/operator-binary/src/superset_controller.rs [hunk 1]
```diff
@@ -490,23 +489,30 @@ fn build_server_rolegroup_statefulset(
         add_authentication_volumes_and_volume_mounts(authentication_class, &mut cb, &mut pb)?;
     }
 
+    let webserver_timeout = node_config
+        .get(&PropertyNameKind::File(SUPERSET_CONFIG_FILENAME.to_string()))
+        .expect("Failed to get the superset config file from node config. It should be set by product config machinery")
+        .get(&SupersetConfigOptions::SupersetWebserverTimeout.to_string())
+        .expect("Failed to get webserver timeout from superset config file. It should be set by product config machinery");
```
Comment (maltesander) [inline rust/operator-binary/src/superset_controller.rs:496]: Do we really wanna stop the operator here? 
Lets assume we have a "bad" product config file (user created without the default for the timeout) and 2 clusters. One with and one without the timeout explicitly configured. One would work the other would make the operator panic. 
Do we want this?

Comment (teozkr) [reply rust/operator-binary/src/superset_controller.rs:496]: Yeah, please don't use `expect()` (or `unwrap()`) for errors that can be caused by user-controlled resources.

Comment (sbernauer) [reply rust/operator-binary/src/superset_controller.rs:496]: My understanding is that the product-config will *always* put in the `SUPERSET_WEBSERVER_TIMEOUT` and even additionally validate that it's there - it's marked as
```
      roles:
        - name: "node"
          required: true
```
So i concluded the error can not be caused by any user input but only by a bug in our product-config machinery.
Was that conclusion wrong? (All tests were green btw)

Comment (teozkr) [reply rust/operator-binary/src/superset_controller.rs:496]: Well, iff the user is using our supplied product-config _and_ our product-config never changes this...

Comment (sbernauer) [reply rust/operator-binary/src/superset_controller.rs]: Switched from the expect to a new Error variant. Hope that's ok

File: rust/operator-binary/src/superset_controller.rs [hunk 2]
```diff
@@ -155,6 +154,10 @@ pub enum Error {
         source: stackable_operator::error::Error,
         authentication_class: ObjectRef<AuthenticationClass>,
     },
+    #[snafu(display("failed to get the superset config file from node config. It should be set by product config machinery"))]
+    MissingSupersetConfigInNodeConfig,
+    #[snafu(display("failed to get webserver timeout from superset config file. It should be set by product config machinery"))]
+    MissingWebServerTimeoutInSupersetConfig,
```
Comment (maltesander) [inline rust/operator-binary/src/superset_controller.rs:160]:
```suggestion
    #[snafu(display(
        "failed to get the {SUPERSET_CONFIG_FILENAME} file from node or product config"
    ))]
    MissingSupersetConfigInNodeConfig,
    #[snafu(display("failed to get {timeout} from {SUPERSET_CONFIG_FILENAME} file. It should be set in the product config or by user input", timeout = SupersetConfigOptions::SupersetWebserverTimeout))]
    MissingWebServerTimeoutInSupersetConfig,
```

**Comment by sbernauer** [comment]
thx!

**Review by maltesander — COMMENTED** [review]
Can we add the timeout in the examples as well as the getting started docs?

**Comment by sbernauer** [comment]
Sure! Added them

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by sbernauer** [comment]
thx!

File: deploy/crd/supersetcluster.crd.yaml [hunk 3]
```diff
@@ -70,9 +70,21 @@ spec:
...
                     properties:
                       rowLimit:
+                        description: Row limit when requesting chart data. Corresponds
+                          to ROW_LIMIT
                         format: int32
                         nullable: true
                         type: integer
+                      webserverTimeout:
+                        description: Maximum number of seconds a Superset request
+                          can take before timing out. This settings effect the maximum
```
Comment (lfrancke) [inline deploy/crd/supersetcluster.crd.yaml:80]: `effect` -> affects

**Comment by lfrancke** [comment]
None of the checkboxes have been ticked (I did one). Could you please complete that?
I'm especially interested if there are tests and documentation. Thank you!

File: deploy/crd/supersetcluster.crd.yaml (hunk 3 above)
Comment (sbernauer) [reply deploy/crd/supersetcluster.crd.yaml:80]: Created https://github.com/stackabletech/superset-operator/pull/261

**Comment by sbernauer** [comment]
We added it to the docs, checked that.
Regarding the tests we tried to come up with an unit or integration test but discarded them as unfeasible

**Pull request closed without merging**